lib/baseFactory.js

```diff
@@ -0,0 +1,135 @@
+'use strict';
+const schema = require('screwdriver-data-schema');
+const hashr = require('screwdriver-hashr');
+const nodeify = require('./nodeify');
+const hoek = require('hoek');
+
+class BaseFactory {
+    /**
+     * Construct a BaseModel object
+     * @method constructor
+     * @param  {String}    modelName            Name of the model to get from data-schema
+     * @param  {Object}    config
+     * @param  {Object}    config.datastore     Object that will perform operations on the datastore
+     */
+    constructor(modelName, config) {
+        this.model = schema.models[modelName];
+        this.table = this.model.tableName;
+        this.datastore = config.datastore;
+    }
+
+    /**
+     * Interface for creating an instance of a Model
+     * @method createClass
+     * @return {Object}
+     */
+    createClass() {
+        throw new Error('must be implemented by extender');
+    }
+
+    /**
+     * Generate the id for the model
+     * @method generateId
+     * @param  {Object}   config Object to generate a hashed ID for
+     * @return {String}          SHA1 unique ID
+     */
+    generateId(config) {
+        const hashObject = {};
+
+        this.model.keys.forEach((keyName) => {
+            hashObject[keyName] = config[keyName];
+        });
+
+        return hashr.sha1(hashObject);
+    }
+
+    /**
+     * Create a Model
+     * @method create
+     * @param  {Object}    config               Config object
+     * @return {Promise}
+     */
+    create(config) {
+        const id = this.generateId(config);
+        const modelConfig = {
+            table: this.table,
+            params: {
+                id,
+                data: config
+            }
+        };
+
+        return nodeify.withContext(this.datastore, 'save', [modelConfig])
+            .then(modelData => this.createClass(hoek.applyToDefaults(modelData, {
+                datastore: this.datastore
+            })));
+    }
+
+    /**
+     * Get a record based on id
+     * @method get
+     * @param  {Mixed}   config    The configuration from which an id is generated or the actual id
+     * @return {Promise}
+     */
+    get(config) {
+        let id = config;
+
+        if (typeof config === 'object' && config.id) {
+            id = config.id;
+        } else if (typeof config === 'object') {
+            id = this.generateId(config);
```

Comment:
We don't always have a generated id for the item we are going to look up, especially when it comes to user models. We can, however, generate an id from the unique keys (e.g. username) and fetch them. Thus saving a step when fetching a model.